File: src/208. Implement Trie (Prefix Tree).js
class Trie {
  /**
   * Initialize your data structure here.
   */
  constructor() {
    this.root = {}
  }
  traverse(word) {
    let current = this.root
    for (let char of word) {
      if (!current[char]) return null
      current = current[char]
    }
    return current
  }
  /**
   * Inserts a word into the trie.
   * @param {string} word
   * @return {void}
   */
  insert(word) {
    let current = this.root
    for (let char of word) {
      if (!current[char]) current[char] = {}
      current = current[char]
    }
    current.isEnd = true
  }
  /**
   * Returns if the word is in the trie.
   * @param {string} word
   * @return {boolean}
   */
  search(word) {
    let node = this.traverse(word)
    return node ? !!node.isEnd : false
  }
  /**
   * Returns if there is any word in the trie that starts with the given prefix.
   * @param {string} prefix
   * @return {boolean}
   */
  startsWith(word) {
    let node = this.traverse(word)
    return !!node
  }
}

/**
 * Your Trie object will be instantiated and called as such:
 * var obj = new Trie()
 * obj.insert(word)
 * var param_2 = obj.search(word)
 * var param_3 = obj.startsWith(prefix)
 */

File: src/208. Implement Trie (Prefix Tree).cpp
class TrieNode {
public:
    bool isLeaf = false;
    TrieNode* children[26];
};

class Trie {
private:
    TrieNode* root;
    TrieNode* getNode (string word) {
        auto node = root;
        for (auto c : word) {
            int index = c - 'a';
            if (node->children[index] == NULL) {
                return NULL;
            }
            node = node->children[index];
        }
        return node;
    }
public:
    Trie() {
        root = new TrieNode();
    }
    
    void insert(string word) {
        auto node = root;
        for (auto c : word) {
            int index = c - 'a';
            if (node->children[index] == NULL) {
                node->children[index] = new TrieNode();
            }
            node = node->children[index];
        }
        node->isLeaf = true;
    }
    
    bool search(string word) {
        auto node = getNode(word);
        return node != NULL && node->isLeaf;
    }
    
    bool startsWith(string prefix) {
        auto node = getNode(prefix);
        return node != NULL;
    }
};

/**
 * Your Trie object will be instantiated and called as such:
 * Trie* obj = new Trie();
 * obj->insert(word);
 * bool param_2 = obj->search(word);
 * bool param_3 = obj->startsWith(prefix);
 */